### ISSUE TITLE
expect.toEquals() class instance check different from jest

### ISSUE THREAD

**Issue by marvinhagemeister** [denoland/deno_std#4244]
**Describe the bug**

Asserting class instances is different in jest's `expect` library compared to ours.

**Steps to Reproduce**

```ts
import { expect } from "https://deno.land/std@0.213.0/expect/mod.ts";

function create() {
  class Person {
    constructor(public readonly name = "deno") {}
  }
  return new Person();
}

Deno.test("should work", () => {
  expect([create()]).toEqual([create()]);
});
```

**Expected behavior**

The test should pass like it does in `jest`.

**Environment**

- OS: macOS 14.2.1
- deno version: 1.40.2
- std version: 0.213.0

**Comment by vinc3re** [comment]
The issue lies in comparing the constructors of the objects of both `Person` objects. I tested it with the following code:

```ts
import { assertEquals } from "std/assert";

function create() {
  class Person {
    constructor(public readonly name = "deno") {}
  }
  return new Person();
}

const a = [create()];
const b = [create()];

Deno.test("should work", () => {
  assertEquals(a, b);
});
```

Using `assertEquals()`.

`assertEquals()` calls `equals()`, so the `equals()` method is our target. In the first call of `equals`, both arrays are passed. The sequence is:

1. Successfully validates that they are objects at [line 54](https://github.com/denoland/deno_std/blob/aad99992653b45d9bd09ccba1f645accb22f5cf4/assert/equal.ts#L54) with `if (a && typeof a === "object" && b && typeof b === "object")`. This is true for both arrays.

2. Goes through all other `if` checks (with `false`) until it reaches [line 104](https://github.com/denoland/deno_std/blob/aad99992653b45d9bd09ccba1f645accb22f5cf4/assert/equal.ts#L104), where `compare()` is recursively called with the `Person` object on each side.

   ```ts
   //...
   const a = [create()];
   const b = [create()];

   console.log(a["0"]);
   console.log(b["0"]);

   //output...

   //Person { name: "deno" }
   //Person { name: "deno" }
   ```

   Before that, of course, the merge of both arrays is done.

   ```ts
   //...
   const a = [create()];
   const b = [create()];

   console.log({ ...a, ...b });

   //output...
   //{ "0": Person { name: "deno" } }
   ```

   And the merging of properties:

   ```ts
   const a = [create()];
   const b = [create()];
   const merge = { ...a, ...b };
   const propMerge = [
     ...Object.getOwnPropertyNames(merge),
     ...Object.getOwnPropertySymbols(merge),
   ];
   console.log(propMerge);
   //output
   //[ "0" ]
   ```

   The key `"0"` is the only key in `propMerge`, and therefore, there will be only one iteration in the `for ... of` loop. So, `a[key]` accesses `{ "0": Person { name: "deno" } }`.

   Then, the second call to the `compare()` method occurs, stacking the first call. However, this time, the call is made with the objects of each array.

3. Again, it goes through all the checks from line 30 until reaching [line 55](https://github.com/denoland/deno_std/blob/aad99992653b45d9bd09ccba1f645accb22f5cf4/assert/equal.ts#L55). I hadn't mentioned the comparison of the constructors of the arrays in the previous call to `compare()`, as this check returns `false`. See:

   ```ts
   const a = [create()];
   const b = [create()];

   function constructorsEqual(a: object, b: object) {
     return (
       a.constructor === b.constructor ||
       (a.constructor === Object && !b.constructor) ||
       (!a.constructor && b.constructor === Object)
     );
   }

   const compareConstructorsOfArray = a && b && !constructorsEqual(a, b);
   console.log(compareConstructorsOfArray);
   //output...

   //false
   ```

   However, this call with the objects generates `true` at line 55. See:

   ```ts
   //...
   const a = [create()];
   const b = [create()];

   function constructorsEqual(a: object, b: object) {
     return (
       a.constructor === b.constructor ||
       (a.constructor === Object && !b.constructor) ||
       (!a.constructor && b.constructor === Object)
     );
   }
   const _a = a["0"];
   const _b = b["0"];
   const compareConstructorsOfPersonObj = _a && _b &&
     !constructorsEqual(_a, _b);
   console.log(_a);
   console.log(_b);
   console.log(compareConstructorsOfPersonObj);
   //output...

   //Person { name: "deno" }
   //Person { name: "deno" }
   //true
   ```

4. Once the condition at line 55 is resolved as `true`, `false` is returned to the previous stacked call at line 56, which had been paused in the `for ... of` loop. The `for ... of` then negates false, `!false`, at [line 104](https://github.com/denoland/deno_std/blob/aad99992653b45d9bd09ccba1f645accb22f5cf4/assert/equal.ts#L104), causing `equal()` to return `false` for `assertEquals()` at [line 32](https://github.com/denoland/deno_std/blob/aad99992653b45d9bd09ccba1f645accb22f5cf4/assert/assert_equals.ts#L32).

This is the entire flow of checks. `equals()` returns false because the constructors of each `Person` object are compared and considered not equal at [line 55](https://github.com/denoland/deno_std/blob/aad99992653b45d9bd09ccba1f645accb22f5cf4/assert/equal.ts#L55). See:

```ts
const _a = a["0"];
const _b = b["0"];
const comparePersonConstructors = constructorsEqual(_a, _b);
console.log(_a);
console.log(_b);
console.log(comparePersonConstructors);

//output

//Person { name: "deno" }
//Person { name: "deno" }
//false
```

I hope this can be helpful.

**Comment by vinc3re** [comment]
Does [this pull request](https://github.com/denoland/deno_std/pull/4246) only get released in version 0.214 of the stdlib?

**Comment by kt3k** [comment]
fixed by #4246

> Does https://github.com/denoland/deno_std/pull/4246 only get released in version 0.214 of the stdlib?

The fix should be available in 0.214.0

```ts
import { expect } from "https://deno.land/std@0.214.0/expect/mod.ts";

function create() {
  class Person {
    constructor(public readonly name = "deno") {}
  }
  return new Person();
}

Deno.test("should work", () => {
  expect([create()]).toEqual([create()]);
});
```
